### PR TITLE
Implemented hiding of outdated bot comments - #1976

### LINKED ISSUE #1976: GitHub Actions: Automate In Progress Issue Updates: Hide Old Comments
### Dependency

- #2032 

### Overview

As developers, we will update the GitHub action job "Add-Update-Label-to-Issues-Weekly" so that duplicate or outdated "update" comments are hidden, in order to improve the readability of the issue history.  

### Details
  If a bot "update" comment is older than `commentByDays` (currently 7), it is either going to be duplicated with another "update" comment (if it is more recent than an assignee comment) or it is outdated (because a more recent assignee comment exists).  In either case the old bot "update" comment should be hidden.  

### Action Items

- [x] The job "Add-Update-Label-to-Issues-Weekly" is part of the workflow `schedule-fri-0700.yml` and the script can be found in [add-label.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/trigger-schedule/add-update-label-weekly/add-label.js).  Understand the code, in particular the function`isTimelineOutdated` in which each timeline event (including comments) is examined
- [x] Add code that hides (as "outdated") bot "update" comments older than `commentByDays`
- [x] Test using your copy of the Project Board, with your PAT, as described in Tip 6 and 7 of [Hack for LA's GitHub Actions](https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions)

### Resources/Instructions
[schedule-fri-0700.yml](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/schedule-fri-0700.yml)
[add-label.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/trigger-schedule/add-update-label-weekly/add-label.js)
Never done GitHub actions? [Start here!](https://docs.github.com/en/actions)
[Hack for LA's GitHub Actions](https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions)
[Events that trigger workflows](https://docs.github.com/en/actions/reference/events-that-trigger-workflows)
[Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions)
[GitHub GraphQL](https://docs.github.com/en/graphql)

### PR DESCRIPTION
**Pull request by iancooperman**
Fixes #1976

### What changes did you make?
  - Added code that hides bot "update" comments older than 7 days.

### Why did you make the changes (we will use this info to test)?
  - To improve readability of issue histories.

### CODE REVIEW

**Comment by roslynwythe** [comment]
@iancooperman Please see the CodeQL alert/annotation above.  Does the log entry indeed depend on a user-provided value?  If you feel that the alert is a false positive, please leave a comment with your reasoning.  Please do not dismiss the alert.   If the alert is legitimate, would you be willing to sanitize the log entry as described in "show more" ?   Thank you for taking up this large and complex issue!

**Comment by roslynwythe** [comment]
- Hi @iancooperman thank you for completing this challenging issue.   I'm currently reviewing another open PR, #6542 which modifies the same file `github-actions/trigger-schedule/add-update-label-weekly/add-label.js` and when that review is complete I will use the same test environment to review this one.

File: github-actions/trigger-schedule/add-update-label-weekly/add-label.js [hunk 1]
```diff
@@ -141,6 +147,58 @@
...
+        res.on('data', (d) => {
+          data += d;
+        });
+        res.on('end', () => {
+          console.log(data);
+        });
+      });
+
+      req.on('error', (error) => {
+        console.error(error);
```
[the remark this answers is not in the document]
Comment (iancooperman) [reply github-actions/trigger-schedule/add-update-label-weekly/add-label.js]: @roslynwythe I’m inclined to consider this as a false positive. The error logging you’re seeing is associated with any potential issues that might arise during the execution of the GraphQL request. This request is solely dependent on the node ID of a comment within an issue. Importantly, this Node ID is not user-generated but is provided directly by GitHub’s REST API. Therefore, in theory, it should always be accurate.

[the remark this answers is not in the document]
Comment (roslynwythe) [reply github-actions/trigger-schedule/add-update-label-weekly/add-label.js]: Thanks @iancooperman for explaining that.  I'll go ahead and dismiss the alert and mark the conversation resolved.

**Comment by roslynwythe** [comment]
@iancooperman Please advise, what are the scopes required for `secrets.HFLA_PROJECT_BOARD_TOKEN` ?  Also prior to merge I assume that I need to generate the token/secret - should that be done from the HackforLABot account?

Also I noticed that workflow_dispatch was added in schedule-fri-0700.yml.  Was that just for testing or is the intention to  retain it is  ?  Is there any risk in allowing that workflow to be manually triggered?

**Comment by iancooperman** [comment]
> @iancooperman Please advise, what are the scopes required for `secrets.HFLA_PROJECT_BOARD_TOKEN` ? Also prior to merge I assume that I need to generate the token/secret - should that be done from the HackforLABot account?
> 
> Also I noticed that workflow_dispatch was added in schedule-fri-0700.yml. Was that just for testing or is the intention to retain it is ? Is there any risk in allowing that workflow to be manually triggered?

@roslynwythe It appears I forgot to revert the name of that token back to what it's called in the canonical repo, as noted in the [docs](https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions#tip-7-using-personal-access-tokens-to-test-in-your-own-project-board:~:text=Now%2C%20when%20you,that%20doesn%27t%20exist.). I'll go ahead and do that. 

As for `workflow_dispatch`, that was purely for testing purposes, but I was hoping we could keep it so the next person having to edit this has one less hoop to jump through when setting up their testing environment. Manually triggering will cause the bot to create a duplicate comment, leading to there being multiple bot comments that aren't minimized (because they're not 7+ days old yet).

**Comment by roslynwythe** [comment]
@iancooperman thank you for the explanation regarding the HACK_FOR_LA_BOT secret.  

@t-will-gillis Ian has suggested retaining `workflow_dispatch` in `schedule-daily-1100.yml` in order to make future testing more convenient.  Do you approve of that change?

**Review by roslynwythe — CHANGES_REQUESTED** [review]
@iancooperman @t-will-gillis - The wiki does refer to secrets.HACK_FOR_LA_BOT however that name does not appear in the list of action secrets for the repository, so I believe it is outdated information.  I apologize for the confusion.  There is a token with a similar name `secrets.HACKFORLA_BOT_PA_TOKEN` which is in use in our workflows for GitHub API requests so I suggest using that one, if Will agrees.

**Comment by iancooperman** [comment]
> @iancooperman @t-will-gillis - The wiki does refer to secrets.HACK_FOR_LA_BOT however that name does not appear in the list of action secrets for the repository, so I believe it is outdated information. I apologize for the confusion. There is a token with a similar name `secrets.HACKFORLA_BOT_PA_TOKEN` which is in use in our workflows for GitHub API requests so I suggest using that one, if Will agrees.

@roslynwythe I've made the replacement. I'll be on a flight during tonight's dev meeting, so unfortunately we won't be able to discuss further then.

**Review by roslynwythe — CHANGES_REQUESTED** [review]
There is an update to the wiki page that describes test tokens at https://github.com/hackforla/website/issues/6537#issuecomment-2041147335.  It appears that `HACKFORLA_BOT_PA_TOKEN` is missing some of the scopes (user and project) that `HACK_FOR_LA_BOT` had.  Would you be willing to retest with a new test token based on the updated documentation?

**Review by roslynwythe — CHANGES_REQUESTED** [review]
@iancooperman The workflow had been working with a previous test token, but I when I created a new test token, with scopes matching HACKFORLA_BOT_PA_TOKEN according to https://github.com/hackforla/website/issues/6537#issuecomment-2041147335, the workflow failed:

image

It appears that the request to minimize the comments requires `repo` scope so I created a new token with that scope, but the workflow is still not hiding comments.  The only errors I see in the logs are about problems deleting labels because the labels cannot be found, but I don't understand that, because I have these labels: `Status: Updated`, `2 weeks inactive` and `To Update !`.   Please confirm exactly what the scopes are in your latest test token.  I will try again in the next day or two.  If you have any ideas please let me know.  The test issue is https://github.com/roslynwythe/website/issues/9 and the log is https://github.com/roslynwythe/website/actions/runs/8772038718/job/24070499006

**Comment by iancooperman** [comment]
@roslynwythe 

Based on the log you linked, it appears that at least some of the comments are being minimized, as the response from the GraphQL API is saying so.
image
But there are so many comments that the secondary rate limits on the API are being run into.
image
The [docs on API best practices](https://docs.github.com/en/rest/using-the-rest-api/best-practices-for-using-the-rest-api?apiVersion=2022-11-28#pause-between-mutative-requests) suggest waiting at least one second between each request. Should I proceed with implementing the recommended delay?

As for the scopes, here's what I have selected:
image

**Comment by roslynwythe** [comment]
@iancooperman I suggest implementing the wait to avoid rate-limiting and also minimizing the scopes of the token so that we can keep privledges to a minimum.

**Comment by t-will-gillis** [comment]
Started reviewing, hope to finish e.o.d. 4/21

**Review by t-will-gillis — COMMENTED** [review]
Hey @iancooperman Great work on this so far! I am posting this review as a comment temporarily, but I can say that my testing has been successful in that for an issue in the "In Progress (Actively Working)" column, comments written by the bot which older than 7 days are closed. I experimented and gave my token only "repo" scope (see below) and for me at least it seems like this is the bare minimum to get this to run. [Log file here](https://github.com/t-will-gillis/website/actions/runs/8791298562) For discussion, we can either add the "repo" scope to an existing token or create a new token. 

I also experimented with how the token gets delivered to the GraphQL to see if I could simplify this - I couldn't. The reason why I was trying is because specifying the token is much easier for the REST API- only one line in the .yml file and no reference needed in the .js file.

One last thing to ask about is I am wondering whether the GraphQL call could/should be pulled out into a separate function or even a separate module instead of being added to `isTimelineOutdated()`. Not saying this is necessary but it might be cleaner. What do you think?

Again, great work so far and thank you for working out the GraphQL. You might have to do a presentation for us! 

<details>
Screenshot 2024-04-22 144440

</details>

**Comment by iancooperman** [comment]
@roslynwythe I can probably have the wait done by this coming Sunday. I won't have much availability until after Thursday though because I have another, rather large, obligation to prepare for.

@t-will-gillis `isTimelineOutdated()` seemed like a convenient place to put all my code because it already loops through every event on an issue, including comments. But I do agree that there are probably cleaner ways to go about it.

**Comment by roslynwythe** [comment]
@t-will-gillis @iancooperman Please advise, should we add the "repo" scope to HACKFORLA_BOT_PA_TOKEN, or should we define a new token for use in this PR?

**Comment by t-will-gillis** [comment]
@roslynwythe @iancooperman I think that the GitHub literature says the the best practice is to only give a token the scopes that it needs and nothing more, which would be an argument to create a new token. I can go ahead and create one if this is what we want to do. For the secret's name `HACKFORLA_REPO_TOKEN` to follow a similar naming pattern as for the other tokens?

**Comment by t-will-gillis** [comment]
@iancooperman 
>@t-will-gillis isTimelineOutdated() seemed like a convenient place to put all my code because it already loops through every event on an issue, including comments. But I do agree that there are probably cleaner ways to go about it.  

Yes, that makes sense since it is already looping through the issue, as you mentioned.

**Comment by roslynwythe** [comment]
> @roslynwythe @iancooperman I think that the GitHub literature says the the best practice is to only give a token the scopes that it needs and nothing more, which would be an argument to create a new token. I can go ahead and create one if this is what we want to do. For the secret's name `HACKFORLA_REPO_TOKEN` to follow a similar naming pattern as for the other tokens?

Yes I agree.  Thanks

File: github-actions/trigger-schedule/add-update-label-weekly/add-label.js [hunk 2]
```diff
@@ -293,4 +309,68 @@
...
+  
+    res.on('data', (d) => {
+      data += d;
+    });
+    res.on('end', () => {
+      console.log(`GraphQL output: ${data}`);
+    });
+  });
+
+  req.on('error', (error) => {
```
[the remark this answers is not in the document]
Comment (iancooperman) [reply github-actions/trigger-schedule/add-update-label-weekly/add-label.js:366]: The is the same warning as before, which was determined to be a false positive.

**Comment by iancooperman** [comment]
@roslynwythe Sorry it's taken so long, but I've added the wait. Go ahead and try it out again when you get the chance.

**Comment by roslynwythe** [comment]
@iancooperman Using a PAT with the "repo" group of scopes in my fork of the repo, all of the "please add update" bot comments older than 7 days were hidden successfully.  Thank you for adding the wait and your patience with changing the token.  Prior to merge we need to create HACKFORLA_REPO_TOKEN.  Please advise @t-will-gillis, if you would like me to create that secret.

**Comment by t-will-gillis** [comment]
Hi @roslynwythe and @iancooperman  because of another automation's requirements, the `HACKFORLA_ADMIN_TOKEN` also needed the "repo" scope. It now has the following scopes so I believe it will work for this automation:
- repo
- admin:org_hook
- write:org

@iancooperman sorry for the delay- I will try to look at this tonight.

**Review by t-will-gillis — COMMENTED** [review]
Hey @iancooperman - Great job. I tested this and for an issue in the "In Progress (actively working)" column, the bot-generated comments older than 7 days are being hidden.

As I mentioned in a previous comment, the `HACKFORLA_ADMIN_TOKEN` is already upgraded to include the "repo" scope because it was needed for a different automation, and so it can be used to replace `HACKFORLA_REPO_TOKEN`.


I noticed a few more things that aren't having a negative impact on the code, but that likely (?) CodeQL will flag at some point. (Many/ most are from before you worked on it) 
- There are a lot of missing semi-colons: eg. lines 15-20, 47-48, 52, 93, 97, 100, and many more  
 
I triggered CodeQL on a fork and it found:  

- Line 43: this variable not being used
- Line 118: variable not being used
- Line 286 `logins` missing a declaration
- Line 278  `assigneesLogins` missing a declaration  

I am not sure why thus far the main repo's CodeQL has not flagged these, but might be worth it to fix these (potential) alerts. 


To recap about the token, these are the scopes for `HACKFORLA_ADMIN_TOKEN`:
- repo ('Full control of private repositories')
- admin:org_hook
- write:org (under 'admin:org') 

Submitting this as a comment for right now. Thanks for your ongoing work...

**Comment by iancooperman** [comment]
@t-will-gillis @roslynwythe I just changed the name of the secret in `schedule-fri-0700.yml`. Anything else I can do?

**Review by t-will-gillis — CHANGES_REQUESTED** [review]
Hi @iancooperman 

- The file `ga-hide-old-comments-1976.code-workspace` is included in the commit - what is this file?
- In `add-label.js`:
  - please confirm whether the variable at line 118 is being used, if not please remove
  - there are extra spaces on line 142. please remove
  - extra line at 189, please remove
  - to prevent future CodeQL alerts, please add semi-colon at end of line 154
  
Thanks

**Comment by iancooperman** [comment]
> Hi @iancooperman
> 
> * The file `ga-hide-old-comments-1976.code-workspace` is included in the commit - what is this file?
> * In `add-label.js`:
>   
>   * please confirm whether the variable at line 118 is being used, if not please remove
>   * there are extra spaces on line 142. please remove
>   * extra line at 189, please remove
>   * to prevent future CodeQL alerts, please add semi-colon at end of line 154
> 
> Thanks

ETA: 5/16 EoD
Availability: 5/16 6pm-9pm

**Comment by iancooperman** [comment]
@t-will-gillis I've made the requested changes in `add-label.js`. `ga-hide-old-comments-1976.code-workspace` was mistakenly pushed when I  was exploring VS Code's workspace settings. It has now been deleted.

**Review by t-will-gillis — APPROVED** [review]
Hey @iancooperman Fantastic- I did one final test and everything seems to be working as intended.

Thanks for all of your work on this!

**Comment by Thinking-Panda** [comment]
Hi @roslynwythe Whenever possible please add you availability and ETA for this PR. Thanks!

**Review by roslynwythe — APPROVED** [review]
Thank you @iancooperman for your work on this complex issue.
